pkg/appfile/validate.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package appfile
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/oam-dev/kubevela/apis/types"
+	"github.com/oam-dev/kubevela/pkg/dsl/model"
+	"github.com/oam-dev/kubevela/pkg/dsl/process"
+)
+
+// ValidateCUESchematicAppfile validates CUE schematic workloads in an Appfile
+func (p *Parser) ValidateCUESchematicAppfile(a *Appfile) error {
+	for _, wl := range a.Workloads {
+		// because helm & kube schematic has no CUE template
+		// it only validates CUE schematic workload
+		if wl.CapabilityCategory != types.CUECategory {
+			continue
+		}
+
+		pCtx, err := newValidationProcessContext(wl, a.Name, a.RevisionName, a.Namespace)
+		if err != nil {
+			return errors.WithMessage(err, "cannot create validationg process context")
+		}
+		for _, tr := range wl.Traits {
+			if tr.CapabilityCategory != types.CUECategory {
+				continue
+			}
+			if err := tr.EvalContext(pCtx); err != nil {
+				return errors.WithMessagef(err, "cannot evaluate trait %q", tr.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func newValidationProcessContext(wl *Workload, appName, revisionName, ns string) (process.Context, error) {
+	// add hook funcs to validate CUE base
+	baseHooks := []process.BaseHook{}
+	// add hook funcs to validate CUE auxiliaries
+	auxiliaryHooks := []process.AuxiliaryHook{validateAuxiliaryNameUnique()}
+
+	if wl.FullTemplate.Reference.Kind == "" {
+		// only validate output has GVK when workload referenced GVK is empty
+		baseHooks = append(baseHooks, validateOutputGVK())
+	}
+
+	pCtx := process.NewContextWithHooks(ns, wl.Name, appName, revisionName, baseHooks, auxiliaryHooks)
+	pCtx.InsertSecrets(wl.OutputSecretName, wl.RequiredSecrets)
+	if len(wl.UserConfigs) > 0 {
+		pCtx.SetConfigs(wl.UserConfigs)
+	}
+	if err := wl.EvalContext(pCtx); err != nil {
+		return nil, errors.Wrapf(err, "evaluate base template app=%s in namespace=%s", appName, ns)
+	}
+	return pCtx, nil
+}
+
+// validateOutputGVK validates whether the output entity in a
+// ComponentDefinition has a GVK after evaluation.
+func validateOutputGVK() process.BaseHook {
```

Comment:
I think this validation is necessary, the `wl.FullTemplate.Reference.Kind` will never be empty as it should be validate and mutate in webhook, this is a missing issue \cc @yangsoon

Reply:
sure, then I will make this validation compulsory.